src/scenes/dashboard/itemDetail/index.js

```diff
@@ -215,53 +218,90 @@ class ItemDetail extends Component {
         });
     }
 
+    // changes the upvote/downvote color to indicate selection, sets the upvote/downvote number before the response comes. if the requisition fails, undo all
     async vote(item, option) {
-        const { props: { profile, token } } = this;
 
+        // uses lodash.cloneDeep to avoid keeping references
+        let originalItem = _.cloneDeep(this.item);
+        const { props: { profile, token } } = this;
+        // user shouldn't vote his own post
         if (profile.id === item.user.id) {
             return;
         }
         if (item.post.votes && item.post.votes[0]) {
             return;
         }
-
-        var response;
-
-        this.setState({ isLoading: true });
-
-        switch (item.entity.type) {
-            case 'post':
-                response = await votePost(this.props.token, item.entity.id, option);
-                break;
-            default:
-                return;
-                break;
+        if (this.state.postingVote) {
+            return;
+        }
+        // uses this state to avoid double clicking, the user is allowed to vote again only when the last request is done
+        this.setState({postingVote: true});
+    
+        if (option === 'upvote') {
+            // user is unsetting his vote
+            if (this.item.option === 1){
+                this.item.option = null;
+                this.item.upvotes_count -=1;
+            } else {
+                // user is setting his vote to up, had the down up checked
+                if (this.item.option === 2){
+                    this.item.option = 1;
+                    this.item.upvotes_count +=1;
+                    this.item.downvotes_count -=1;
+                } else {
+                    // didnt have any option checked
+                    this.item.option = 1;
+                    this.item.upvotes_count +=1;
+                }
+            }
```

Comment:
why not
```js
switch (this.item.option) { 
    case 1: break;
    case 2: break;
    default: ;
}
```
instead of condition in condition?

Reply:
Personally I'm not a fan of switch when there are only a few cases and,  in this case, it would have only 2 "cases" and more condition inside each "case". the logic has 3 level of conditions that cannot be avoided easily. It feels to me it's more organized to use condition inside condition, since there are 2 main conditions and 3 level of conditions.

Reply:
at least `if {} else if {} else {}`, it's more readable